[Doc] Add doc for the coordinated commits writer feature (#3203)
delta-io/delta: docs/source/delta-managed-commit.md

---
description: Learn about Delta Managed Commits.
---

# Delta Managed Commits

> .. warning:: This feature is available in preview in <Delta> 4.0. Since this feature is still in preview, it can undergo major changes. Furthermore, a table which has this preview feature enabled cannot be written to by future Delta releases until

Reviewer: This is not rendering as expected in the github markdown viewer.

Author: It should render correctly in the docs: 
<img width="1561" alt="image" src="https://github.com/delta-io/delta/assets/125324092/7278529b-3910-4f67-9fb5-1868ede12b2c">

this feature is manually dropped from the table.

> The [Managed Commit](https://github.com/delta-io/delta/issues/2598) writer table feature defines a general protocol which can be extended to define custom commit protocols. The feature enables clients to designate "commit owners" for tables, which means that every commit to the table must be managed by that commit owner.

Reviewer: ```suggestion
[Managed Commit](https://github.com/delta-io/delta/issues/2598) is a new commit protocol which makes the commit process more flexible and pluggable by giving control of the commit to an external commit owner. Each managed commit table has a designated "commit owner" and all the commits to the table must go via it.

> The presence of a commit owner which is responsible for all commits to a table makes multi-cluster and multi-engine writes to tables safe. The commit owner for a table can be discovered by reading the table through the filesystem.

Reviewer: ```suggestion
The presence of a commit owner which is responsible for all commits to a table makes multi-cluster and multi-engine writes to tables safe.
```

> 

Reviewer: Should we add a compatibility section?

  ## Compatibility
  Managed Commit is a writer table feature.
  - That means only clients which understands this feature could write to a managed commit table. 
  - Older clients which do not understand this table feature could still read a managed commit table. Note that the read may give stale results depending on [table's commit owner backfill policy](#hyperLinkToBackfillSectionInProtocolRFC.md).
  ```

# DynamoDB Commit Owner

> <Delta> 4.0 also introduces a reference Commit Owner implementation which uses DynamoDB as the backend.

Reviewer: Should this be 4.0-preview?

> <Delta> 4.0 also introduces a reference Commit Owner implementation which uses DynamoDB as the backend.

Reviewer: ```suggestion
<Delta> 4.0-preview also introduces a DynamoDB backed Commit Owner implementation.
```

## Quickstart Guide

> #. Create the DynamoDB table

Reviewer: ```suggestion
# Create the DynamoDB table
```

> #. Create the DynamoDB table

Reviewer: What does `#.` do? Doesn't render in github UI.

Author: Changed to 1. to create an ordered list.

### 2. Launch a spark shell with the right packages:

```bash
bin/spark-shell \
--packages io.delta:delta-spark_2.13:4.0.0,org.apache.hadoop:hadoop-aws:3.3.4 \
--repositories https://oss.sonatype.org/content/repositories/iodelta-1147 \
--conf spark.sql.extensions=io.delta.sql.DeltaSparkSessionExtension \
--conf spark.sql.catalog.spark_catalog=org.apache.spark.sql.delta.catalog.DeltaCatalog \
--conf spark.databricks.delta.managedCommits.commitOwner.ddb.awsCredentialsProviderName=<credentialsProviderName>
```

`<credentialsProviderName>` must be the fully qualified class name of the [AWS Credentials Provider](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/AWSCredentialsProvider.html) (e.g. `com.amazonaws.auth.profile.ProfileCredentialsProvider`) which must be used for authenticating with the desired DynamoDB instance.

> ### 3. Create a table with DynamoDB as the commit owner by running the following command:

Reviewer: This is a title o a subsection.

```suggestion
### 3. Creating a managed-commit table with DynamoDB as the commit owner
```

Author: In that case:
"### 1. Create the DynamoDB table" --> "### 1. Creating the DynamoDB table"

"### 2. Launch a spark shell with the right packages:" --> "### 2. Launching the spark shell"
?

```sql
CREATE TABLE <table_name> (id string) USING DELTA

> TBLPROPERTIES ('delta.managedCommits.commitOwner-preview' = 'dynamodb', 'delta.managedCommits.commitOwnerConf-preview' = '{\"managedCommitsTableName\": \"<dynamodb_table_name>\",\"dynamoDBEndpoint\": \"<dynamodb_region_endpoint>\"}');

Reviewer: `managedCommitsTableName` should be `dynamoDBTableName` .
